---
title: "IdentificationOfDEgenes"
output: html_document
---

Code used to import and handle count data generated by the prepDE.py script. 
Initial testing of this code can be found in the later half of tuxedoRNAseq_markdownOut.Rmd




EdgeR
================================================================================================================================================================
================================================================================================================================================================

```{r}
rm(list=ls())

source("https://bioconductor.org/biocLite.R")
biocLite()
library(edgeR)
biocLite("ggplot2")
biocLite("stringr")

library(ggplot2)
library(stringr)
library(ggplot2)
library(matrixStats)
library(pheatmap)

countData <- as.matrix(read.csv("~/hisatAnalysis/stringtie_prepDE_output/stringtiePrepDE_geneCountMatrix.csv", row.names="gene_id"))
group <- factor(c(rep("A",3),rep("B",3),rep("C",3),rep("D",3),rep("E",3),rep("F",3)))
sampleInformation <- read.csv("~/hisatAnalysis/ballgownPhenodata.csv")

# Filter genes not expressed:
y <- DGEList(counts = countData, group=group)
keep <- rowSums(cpm(y)>2) >= 3
table(keep)
y <- y[keep, , keep.lib.sizes=FALSE]
y <- calcNormFactors(y)
# The calcNormFactors function normalizes for RNA composition by finding a set of scaling factors for the library sizes that minimize the log-fold changes between the samples for most genes.
## 

# Perform quasi-likelihood F-tests: (These are noted in the user guide to be the correct test to use as it uses stricter error rate control by accounting for the uncertainty in dispersion estimates)
## Refer to EdgeR user's guide Chapter 3 (Specific experimental designs)
## GLM approach (requires a design matrix, allows more general comparisons to be made)

design <- model.matrix(~0+group, data=y$samples)
colnames(design) <- levels(y$samples$group)
design


# Estimate dispersion:
yDisp <- estimateDisp(y, design)

# Plot dispersion:
plotBCV(yDisp)
pdf("/home/mcity599/Documents/tuxedoRNASeq/qualityPlots/BCV.pdf") # starts writing a PDF to file
plotBCV(yDisp)                 # makes the actual plot
           # closes the PDF file
while (!is.null(dev.list()))  
  dev.off()

## Alternative dispersion estimate (stepwise, not recommended):
y2 <- estimateGLMCommonDisp(y,design)
y2 <- estimateGLMTrendedDisp(y2,design)
y2 <- estimateGLMTagwiseDisp(y2,design)

# However, when looking at the two plots produced, it looks like the stepwise dispersion estimate has fitted a more accurate trend line. 
plotBCV(y2)
pdf("/home/mcity599/Documents/tuxedoRNASeq/qualityPlots/BCV_stepwiseDispersionEstimate.pdf") # starts writing a PDF to file
plotBCV(y2)                 # makes the actual plot
           # closes the PDF file
while (!is.null(dev.list()))  
  dev.off()

# Cluster samples using PCA (DESeq2) or MDS (EdgeR)

# In EdgeR, use plotMDS instead of PCA. 
## This function is a variation on the usual multdimensional scaling (or principle coordinate) plot - not a Principle component analysis.
## Dimension 1 is the direction that best separates the samples, without regard to whether they are treatments or replicates. Dimension 2 is the next best direction, uncorrelated with the first, that separates the samples.
plotMDS(y, cex=0.4, main="edgeR MDS Plot")
pdf("/home/mcity599/Documents/tuxedoRNASeq/qualityPlots/MDS.pdf") # starts writing a PDF to file
plotMDS(y, cex=0.4, main="edgeR MDS Plot")               # makes the actual plot
           # closes the PDF file
while (!is.null(dev.list()))  
  dev.off()



# Calculate fit:
fit <- glmQLFit(y2, design)

# Use the makeContrasts function to set up potential tests between the three group-pairs:
my.contrasts <- makeContrasts(BvsA=B-A, DvsC=D-C, FvsE=F-E, levels=design)

# Run tests:
qlf.BvsA <- glmQLFTest(fit, contrast=my.contrasts[,"BvsA"])
# topTags(qlf.BvsA)
qlf.DvsC <- glmQLFTest(fit, contrast=my.contrasts[,"DvsC"])
# topTags(qlf.DvsC)
qlf.FvsE <- glmQLFTest(fit, contrast=my.contrasts[,"FvsE"])
# topTags(qlf.FvsE)

plotMD(qlf.BvsA,main="WT vs NR") |abline(h=c(-1, 1), col="blue")
plotMD(qlf.DvsC) |abline(h=c(-1, 1), col="blue")
plotMD(qlf.FvsE) |abline(h=c(-1, 1), col="blue")

# Put the different experiments into a list which can be treated as one unit for further manipulations:
qlf_list <- list(qlf.BvsA,qlf.DvsC,qlf.FvsE)
names(qlf_list) <- c("qlf.BvsA","qlf.DvsC","qlf.FvsE")
rm(qlf.BvsA)
rm(qlf.DvsC)
rm(qlf.FvsE)

# Identify genes statistically significantly different between the different groups: 
qlfSigPTags <- lapply(1:length(qlf_list),function(x) topTags(qlf_list[[x]], n=Inf, adjust.method = "fdr", sort.by ="PValue", p.value = 0.05))
unlist(lapply(1:3,function(x) dim(qlfSigPTags[[x]]))) # Prints out the number of statistically different genes

# Extract only logFC (log2FC) > 1 or < -1
qlfSigFoldTags <- lapply(1:length(qlfSigPTags),function(i) qlfSigPTags[[i]]$table)
names(qlfSigFoldTags) <- c("qlf.BvsA","qlf.DvsC","qlf.FvsE")

qlfSigFoldTagsDown <- lapply(1:length(qlfSigPTags),function(t) qlfSigFoldTags[[t]][order(qlfSigFoldTags[[t]]$logFC),][qlfSigFoldTags[[t]][order(qlfSigFoldTags[[t]]$logFC),1] < -1,])

qlfSigFoldTagsUp <- lapply(1:length(qlfSigPTags),function(t) qlfSigFoldTags[[t]][order(qlfSigFoldTags[[t]]$logFC),][qlfSigFoldTags[[t]][order(qlfSigFoldTags[[t]]$logFC),1] >1,])

save(qlfSigFoldTagsDown,file='~/Documents/tuxedoRNASeq/geneAnalysis/EdgeR/qlfSigFoldTagsDown.RData')
save(qlfSigFoldTagsUp,file='~/Documents/tuxedoRNASeq/geneAnalysis/EdgeR/qlfSigFoldTagsUp.RData')

# View number of DE genes:
unlist(lapply(1:3,function(x) length(rownames(qlfSigFoldTagsDown[[x]])))) # 105  803 5117 tags downregulated in the first group (BvsA) of each of the three expts
unlist(lapply(1:3,function(x) length(rownames(qlfSigFoldTagsUp[[x]])))) # 183 2770 5588 tags upregulated in the frist group in each of the three expts

# Extract mstrg tags, convert
tagsDown <- lapply(1:3,function(x) rownames(qlfSigFoldTagsDown[[x]]))
tagsUp <- lapply(1:3,function(x) rownames(qlfSigFoldTagsUp[[x]]))
# Convert:
load("~/Documents/tuxedoRNASeq/objectsTuxedoDESeq/mstrgToGene.RData")
EdgeR_downRegGenes <- lapply(1:3,function(b) mstrgToGene[na.omit(match(tagsDown[[b]],mstrgToGene[,1])),2])
EdgeR_upRegGenes <- lapply(1:3,function(b) mstrgToGene[na.omit(match(tagsUp[[b]],mstrgToGene[,1])),2])

# Remember that some genes are labelled with geneXXXX instead of mstrg tag. These can be converted using the geneId converter from Xenbase:
Xlaevis9.2 <- read.delim("~/Documents/xenbaseFiles/Xlaevis9.2.geneID-geneName_Conversion")
EdgeR_downRegGenes[[2]] <- c(EdgeR_downRegGenes[[2]],as.vector(Xlaevis9.2[as.vector(na.omit(match(rownames(qlfSigFoldTagsDown[[2]])[grep("gene",rownames(qlfSigFoldTagsDown[[2]]))],Xlaevis9.2[,1]))),2]))
EdgeR_downRegGenes[[3]] <- c(EdgeR_downRegGenes[[3]],as.vector(Xlaevis9.2[as.vector(na.omit(match(rownames(qlfSigFoldTagsDown[[3]])[grep("gene",rownames(qlfSigFoldTagsDown[[3]]))],Xlaevis9.2[,1]))),2]))

EdgeR_upRegGenes[[1]] <- c(EdgeR_upRegGenes[[1]],as.vector(Xlaevis9.2[as.vector(na.omit(match(rownames(qlfSigFoldTagsUp[[1]])[grep("gene",rownames(qlfSigFoldTagsUp[[1]]))],Xlaevis9.2[,1]))),2]))
EdgeR_upRegGenes[[2]] <- c(EdgeR_upRegGenes[[2]],as.vector(Xlaevis9.2[as.vector(na.omit(match(rownames(qlfSigFoldTagsUp[[2]])[grep("gene",rownames(qlfSigFoldTagsUp[[2]]))],Xlaevis9.2[,1]))),2]))
EdgeR_upRegGenes[[3]] <- c(EdgeR_upRegGenes[[3]],as.vector(Xlaevis9.2[as.vector(na.omit(match(rownames(qlfSigFoldTagsUp[[3]])[grep("gene",rownames(qlfSigFoldTagsUp[[3]]))],Xlaevis9.2[,1]))),2]))


# Save objects:
save(EdgeR_upRegGenes,file='~/Documents/tuxedoRNASeq/geneAnalysis/EdgeR/EdgeR_upRegGenes.RData')
save(EdgeR_downRegGenes,file='~/Documents/tuxedoRNASeq/geneAnalysis/EdgeR/EdgeR_downRegGenes.RData')


write.table(EdgeR_upRegGenes[[1]],"~/Documents/tuxedoRNASeq/geneAnalysis/EdgeR/upregulated in NR compared to WT.csv", row.names=F, col.names = F)
write.table(EdgeR_upRegGenes[[2]],"~/Documents/tuxedoRNASeq/geneAnalysis/EdgeR/upregulated in PNT compared to Tbud.csv", row.names=F, col.names = F)
write.table(EdgeR_upRegGenes[[1]],"~/Documents/tuxedoRNASeq/geneAnalysis/EdgeR/upregulated in InjCap compared to ConCap.csv", row.names=F, col.names = F)

write.table(EdgeR_downRegGenes[[1]],"~/Documents/tuxedoRNASeq/geneAnalysis/EdgeR/upregulated in WT compared to NR.csv", row.names=F, col.names = F)
write.table(EdgeR_downRegGenes[[2]],"~/Documents/tuxedoRNASeq/geneAnalysis/EdgeR/upregulated in Tbud compared to PNT.csv", row.names=F, col.names = F)
write.table(EdgeR_downRegGenes[[3]],"~/Documents/tuxedoRNASeq/geneAnalysis/EdgeR/upregulated in ConCap compared to InjCap.csv", row.names=F, col.names = F)

```




DESeq2
================================================================================================================================================================
================================================================================================================================================================

```{r}
rm(list=ls())

library(DESeq2)

# Read in count data:
countData <- as.matrix(read.csv("~/hisatAnalysis/stringtie_prepDE_output/stringtiePrepDE_geneCountMatrix.csv", row.names="gene_id"))
# Load sample information:
sampleInformation <- read.csv("~/hisatAnalysis/ballgownPhenodata.csv")
# Check that sample ID match between countData and sampleInformation
all(sampleInformation$ids %in% colnames(countData))

# Estimate new counts and variance
dds <- DESeqDataSetFromMatrix(countData = countData, colData = sampleInformation, design = ~ TissueType )
keep <- rowSums(cpm(counts(dds))>2) >=3
table(keep)
dds <- dds[keep,]
```

```{r}
#Look at the PCA
plotPCA(rlog(dds,blind=T), intgroup="TissueType")+theme_bw()
pdf("/home/mcity599/Documents/tuxedoRNASeq/qualityPlots/PCA_DESeq2.pdf")
plotPCA(rlog(dds,blind=T), intgroup="TissueType")+theme_bw()          
while (!is.null(dev.list()))  
  dev.off()
```

```{r}
pcaData <- plotPCA(rlog(dds,blind=T), intgroup=c("TissueType"), returnData=TRUE)
plot(pcaData$PC1, pcaData$PC2,col="white") | text(pcaData$PC1, pcaData$PC2,colnames(dds),cex=1)
pdf("/home/mcity599/Documents/tuxedoRNASeq/qualityPlots/pcaData_DESeq2.pdf")
plot(pcaData$PC1, pcaData$PC2,col="white") | text(pcaData$PC1, pcaData$PC2,colnames(dds),cex=1)
while (!is.null(dev.list()))
  dev.off()
```


Differential expression
```{r}
dds <- DESeq(dds)
# estimating size factors
# estimating dispersions
# gene-wise dispersion estimates
# mean-dispersion relationship
# final dispersion estimates
# fitting model and testing

# Visualise dispersion:
plotDispEsts( dds)
pdf("/home/mcity599/Documents/tuxedoRNASeq/qualityPlots/dispersion_DESeq2.pdf")
plotDispEsts( dds)
while (!is.null(dev.list()))
  dev.off()
```

Identifying DE
```{r}
# Generate results tables
WTN1res <- results(dds, contrast=c("TissueType", "Wtblast", "N1nr"))
TbudPNTres <- results(dds, contrast=c("TissueType", "Tbud", "PNT"))
ConCapInjCapres <- results(dds, contrast=c("TissueType", "ConCap", "InjCap"))
# Apply lfcShrink
WTN1res.lfc <- lfcShrink(dds, contrast=c("TissueType", "Wtblast", "N1nr"), res = WTN1res)
TbudPNTres.lfc <- lfcShrink(dds, contrast=c("TissueType", "Tbud", "PNT"), res = TbudPNTres)
ConCapInjCapres.lfc <- lfcShrink(dds, contrast=c("TissueType", "ConCap", "InjCap"), res = ConCapInjCapres)

ShrunkResults <- list(WTN1res.lfc, TbudPNTres.lfc, ConCapInjCapres.lfc)
names(ShrunkResults) <- c("WTN1res.lfc","TbudPNTres.lfc","ConCapInjCapres.lfc")

# Remove rows where padj = na
ShrunkResults <- lapply(1:length(ShrunkResults),function(x) ShrunkResults[[x]][!is.na(ShrunkResults[[x]]$padj),])

# Order the results table, retain padj < 0.05
shrunkPadjOrd <- lapply(1:length(ShrunkResults),function(x) ShrunkResults[[x]][order(ShrunkResults[[x]]$padj),])
statisticallySignificantShrunkPadj <- lapply(1:length(ShrunkResults),function(x) shrunkPadjOrd[[x]][as.vector(na.omit(shrunkPadjOrd[[x]]$padj < 0.05)),])
statisticallySignificantShrunkPadj[[1]]

# Order by log2FC, retain log2FC >1 or < -1
logOrd <- lapply(1:length(shrunkPadjOrd),function(x) statisticallySignificantShrunkPadj[[x]][order(statisticallySignificantShrunkPadj[[x]]$log2FoldChange),])
DEmstrgTagsNeg <- lapply(1:length(logOrd),function(x) logOrd[[x]][as.vector(na.omit(logOrd[[x]]$log2FoldChange < -1)),])
DEmstrgTagsPos <- lapply(1:length(logOrd),function(x) logOrd[[x]][as.vector(na.omit(logOrd[[x]]$log2FoldChange > 1)),])

# Tidy up:
names(DEmstrgTagsPos) <- c("WtVsN1nr","TbudVsPNT","ConCapVsInjCap")
names(DEmstrgTagsNeg) <- c("WtVsN1nr","TbudVsPNT","ConCapVsInjCap")

save(DEmstrgTagsNeg,file='~/Documents/tuxedoRNASeq/geneAnalysis/DESeq2/DEmstrgTagsNeg.RData')
save(DEmstrgTagsPos,file='~/Documents/tuxedoRNASeq/geneAnalysis/DESeq2/DEmstrgTagsPos.RData')
save(ShrunkResults,file="~/Documents/tuxedoRNASeq/geneAnalysis/DESeq2/ShrunkResults.RData")

## Due to a quirk of how I oriented my contrast in EdgeR (BvsA) compared to DESeq2 (AvsB), probes which here are "Neg" are downregulated in A(WT,TBud,ConCap) or upregulated in NR, PNT, InjCap.
## I.e., DEmstrgTagsPos are upregulated in WT, Tbud, ConCap
```


Make heatmaps of DE gene data
```{r}
biocLite("pheatmap")
library(pheatmap)

rld <- rlog(dds,blind = F)
###regularized logarithm or rlog normalisation which is suited for data visualisation (unlike the one we used for actually testing). It incorporates a prior on the sample differences (Love, Huber, and Anders 2014). , data on the log2 scale which has been normalized with respect to library size or other normalization factors. Difference to the average expression of every differentially expressed gene. the counts are transformed using regularized logarithm (Love, Huber, and Anders 2014).
# rlog is a way to turn RNA-seq data (in which gene variance is often related to the mean) to (approximately) homoskedastic data (variance is not related to the mean)

matNegGenes <- lapply(1:length(DEmstrgTagsNeg),function(x) assay(rld)[which(rownames(assay(rld)) %in%rownames(DEmstrgTagsNeg[[x]])),])
matNegGenes <- lapply(1:3,function(x) matNegGenes[[x]] - rowMeans(matNegGenes[[x]]))
matPosGenes <- lapply(1:length(DEmstrgTagsPos),function(x) assay(rld)[which(rownames(assay(rld)) %in%rownames(DEmstrgTagsPos[[x]])),])
matPosGenes <- lapply(1:3,function(x) matPosGenes[[x]] - rowMeans(matPosGenes[[x]]))
names(matNegGenes) <- c("UpregulatedInNR","UpregulatedInPNT","UpregulatedInInjCap")
names(matPosGenes) <- c("UpregulatedInWT","UpregulatedInTbud","UpregulatedInConCap")

# Annotation (experimental group, tissue sample type)
groupAnnotation = data.frame(
                    Tissue = factor(c(rep("WT",3),rep("NR",3),rep("Tbud",3),rep("PNT",3),rep("ConCap",3),rep("InjCap",3))), Experiment = factor(c(rep("1",6),rep("2",6),rep("3",6)))
                )
rownames(groupAnnotation) = sampleInformation$ids
groupAnnotation


## Heatmap options: 
# Can reduce cellheight and size to make image fit to one page, but lose ability to distinguish rownames.
# cluster_col = T forces clustering of samples

pdf("/home/mcity599/Documents/tuxedoRNASeq/geneAnalysis/DESeq2/DEgene_heatmaps/DEgeneHeatmaps_WTvsNR.pdf")
pheatmap(rbind(matNegGenes[[1]],matPosGenes[[1]]),annotation_col = groupAnnotation, show_rownames = F,border_color="lightgrey",cluster_row=T,cluster_col=T,fontsize=6,main='DE genes: WT vs NR') 
while (!is.null(dev.list()))
  dev.off()
# Tested with cluster_col=T, result: samples all cluster within groups (e.g., ACB one cluster, DEF another etc.,) - WT and NR are next to each other and, with PNT, form a cluster which contains half the samples. Within the other half, it is ConCap that forms the outgroup, with InjCap and Tbud grouping.  

pdf("/home/mcity599/Documents/tuxedoRNASeq/geneAnalysis/DESeq2/DEgene_heatmaps/DEgeneHeatmaps_TbudVsPNT.pdf")
pheatmap(rbind(matNegGenes[[2]],matPosGenes[[2]]),annotation_col = groupAnnotation, show_rownames = F,border_color="lightgrey",cluster_row=T,cluster_col=T,fontsize=6,main='DE genes: Tbud vs PNT')
while (!is.null(dev.list()))
  dev.off()

pdf("/home/mcity599/Documents/tuxedoRNASeq/geneAnalysis/DESeq2/DEgene_heatmaps/DEgeneHeatmaps_ConCapVsInjCap.pdf")
pheatmap(rbind(matNegGenes[[3]],matPosGenes[[3]]),annotation_col = groupAnnotation, show_rownames = F,border_color="lightgrey",cluster_row=T,cluster_col=T,fontsize=6,main='DE genes: ConCap vs InjCap')
while (!is.null(dev.list()))
  dev.off()
```


Extract gene names for the DESeq2 DE mstrg tags
```{r}
load("~/Documents/tuxedoRNASeq/objectsTuxedoDESeq/mstrgToGene.RData")

DESeq2_negRegGenes <- lapply(1:3,function(b) mstrgToGene[na.omit(match(rownames(DEmstrgTagsNeg[[b]]),mstrgToGene[,1])),2])
DESeq2_posRegGenes <- lapply(1:3,function(b) mstrgToGene[na.omit(match(rownames(DEmstrgTagsPos[[b]]),mstrgToGene[,1])),2])
names(DESeq2_negRegGenes) <- names(DEmstrgTagsNeg)
names(DESeq2_posRegGenes) <- names(DEmstrgTagsPos)

# Remember that some genes are labelled with geneXXXX instead of mstrg tag. These can be converted using the geneId converter from Xenbase:
Xlaevis9.2 <- read.delim("~/Documents/xenbaseFiles/Xlaevis9.2.geneID-geneName_Conversion")

DESeq2_negRegGenes <- lapply(1:3,function(x) c(DESeq2_negRegGenes[[x]],as.vector(Xlaevis9.2[as.vector(na.omit(match(rownames(DEmstrgTagsNeg[[x]])[grep("gene",rownames(DEmstrgTagsNeg[[x]]))],Xlaevis9.2[,1]))),2])))
DESeq2_posRegGenes <- lapply(1:3,function(x) c(DESeq2_posRegGenes[[x]],as.vector(Xlaevis9.2[as.vector(na.omit(match(rownames(DEmstrgTagsPos[[x]])[grep("gene",rownames(DEmstrgTagsPos[[x]]))],Xlaevis9.2[,1]))),2])))


# Save objects
save(DESeq2_negRegGenes,file='~/Documents/tuxedoRNASeq/geneAnalysis/DESeq2/DESeq2_negRegGenes.RData')
save(DESeq2_posRegGenes,file='~/Documents/tuxedoRNASeq/geneAnalysis/DESeq2/DESeq2_posRegGenes.RData')

write.table(DESeq2_negRegGenes[[1]],"~/Documents/tuxedoRNASeq/geneAnalysis/DESeq2/upregulated in NR compared to WT.csv",row.names = F, col.names = F)
write.table(DESeq2_negRegGenes[[2]],"~/Documents/tuxedoRNASeq/geneAnalysis/DESeq2/upregulated in PNT compared to TBud.csv",row.names = F, col.names = F)
write.table(DESeq2_negRegGenes[[3]],"~/Documents/tuxedoRNASeq/geneAnalysis/DESeq2/upregulated in InjCap compared to ConCap.csv",row.names = F, col.names = F)

write.table(DESeq2_posRegGenes[[1]],"~/Documents/tuxedoRNASeq/geneAnalysis/DESeq2/upregulated in WT compared to NR.csv",row.names = F, col.names = F)
write.table(DESeq2_posRegGenes[[2]],"~/Documents/tuxedoRNASeq/geneAnalysis/DESeq2/upregulated in Tbud compared to PNT.csv",row.names = F, col.names = F)
write.table(DESeq2_posRegGenes[[3]],"~/Documents/tuxedoRNASeq/geneAnalysis/DESeq2/upregulated in ConCap compared to InjCap.csv",row.names = F, col.names = F)
```



Produce an expanded range of improved pheatmaps
====================================================================
- Pool DE genes 
- Intersection of DE genes between groups
- Attach gene names to rows when n rows is small
```{r}
# Produce new rlog - transformed objects which have gene names attached. 
rld2 <- rlog(dds,blind = F)
names(rld2) <- mstrgToGene[match(names(rld2),mstrgToGene[,1]),2]
save(rld2,file="~/Documents/tuxedoRNASeq/geneAnalysis/DESeq2/rld2.RData")
rld_negGenes <- lapply(1:length(DESeq2_negRegGenes),function(x) assay(rld2)[which(rownames(assay(rld2)) %in%DESeq2_negRegGenes[[x]]),])
rld_negGenes <- lapply(1:3,function(x) rld_negGenes[[x]] - rowMeans(rld_negGenes[[x]]))

rld_posGenes <- lapply(1:length(DESeq2_posRegGenes),function(x) assay(rld2)[which(rownames(assay(rld2)) %in%DESeq2_posRegGenes[[x]]),])
rld_posGenes <- lapply(1:3,function(x) rld_posGenes[[x]] - rowMeans(rld_posGenes[[x]]))

names(rld_negGenes) <- c("UpregulatedInNR","UpregulatedInPNT","UpregulatedInInjCap")
names(rld_posGenes) <- c("UpregulatedInWT","UpregulatedInTbud","UpregulatedInConCap")

# Annotation (experimental group, tissue sample type)
groupAnnotation = data.frame(
                    Tissue = factor(c(rep("WT",3),rep("NR",3),rep("Tbud",3),rep("PNT",3),rep("ConCap",3),rep("InjCap",3))), Experiment = factor(c(rep("1",6),rep("2",6),rep("3",6)))
                )
rownames(groupAnnotation) = sampleInformation$ids
groupAnnotation

# Create groups of genes:
Up_WtTbudInCap <- c(unique(rownames(rld_posGenes[[1]])),unique(rownames(rld_posGenes[[2]])),unique(rownames(rld_negGenes[[3]])))
Up_NrPntConCap <- c(unique(rownames(rld_negGenes[[1]])),unique(rownames(rld_negGenes[[2]])),unique(rownames(rld_posGenes[[3]])))

table(table(Up_WtTbudInCap))[3] # This code can be used to provide a length command

gC <- names(table(Up_WtTbudInCap)[order(table(Up_WtTbudInCap),decreasing = T)[1:table(table(Up_WtTbudInCap))[3]]])
sampleOrd <- c(1,2,3,4,5,6,7,8,9,10,11,12,16,17,18,13,14,15) # Places samples in order of WT/NR, Tbud/PNT, InjCap/ConCap (i.e., regen classes on left)
pdf("/home/mcity599/Documents/tuxedoRNASeq/geneAnalysis/DESeq2/DEgene_heatmaps/DEgeneHeatmaps_upregulatedWtTbudInjCap.pdf")
pheatmap(assay(rld2)[which(rownames(assay(rld2)) %in% gC),sampleOrd],annotation_col = groupAnnotation, show_rownames = T,border_color="lightgrey",cluster_row=T,cluster_col=F,fontsize=6,main='Upregulated in WT, Tbud, and InjCap')
while (!is.null(dev.list()))
  dev.off()

gC <- names(table(Up_NrPntConCap)[order(table(Up_NrPntConCap),decreasing = T)[1:table(table(Up_NrPntConCap))[3]]])
pdf("/home/mcity599/Documents/tuxedoRNASeq/geneAnalysis/DESeq2/DEgene_heatmaps/DEgeneHeatmaps_upregulatedNrPNTConCap.pdf")
pheatmap(assay(rld2)[which(rownames(assay(rld2)) %in% gC),sampleOrd],annotation_col = groupAnnotation, show_rownames = T,border_color="lightgrey",cluster_row=T,cluster_col=F,fontsize=6,main='Upregulated in Nr, PNT, and ConCap')
while (!is.null(dev.list()))
  dev.off()


pdf("/home/mcity599/Documents/tuxedoRNASeq/geneAnalysis/DESeq2/DEgene_heatmaps/DEgeneHeatmaps_upregulatedInjCap.pdf")
pheatmap(assay(rld2)[which(rownames(assay(rld2)) %in% rownames(rld_negGenes[[3]])),sampleOrd],annotation_col = groupAnnotation, show_rownames = F,border_color="lightgrey",cluster_row=T,cluster_col=F,fontsize=6,main='Upregulated in InjCap')
while (!is.null(dev.list()))
  dev.off()
```


Using heatmaps to visualise sample distance:
```{r}
sampleDists <- dist( t( assay(rld2) ) )
sampleDists

library("RColorBrewer")

sampleDistMatrix <- as.matrix( sampleDists )
rownames(sampleDistMatrix) <- paste( rld2$TissueType, rld2$ids, sep="-" )
colnames(sampleDistMatrix) <- NULL
colors <- colorRampPalette( rev(brewer.pal(9, "Blues")) )(255)
pheatmap(sampleDistMatrix,
         clustering_distance_rows=sampleDists,
         clustering_distance_cols=sampleDists,
         col=colors)
# Save plot:
pdf("/home/mcity599/Documents/tuxedoRNASeq/qualityPlots/sampleDistanceHeatmap_DESeq2.pdf")
pheatmap(sampleDistMatrix,
         clustering_distance_rows=sampleDists,
         clustering_distance_cols=sampleDists,
         col=colors)
while (!is.null(dev.list()))
  dev.off()

```

















A note on differences between EdgeR and DESeq2, with a specific example:

Both EdgeR and DESeq2 were used to identify DE genes on the basis of gene-level counts obtained through the PrepDE.py script. 
The majority of genes identified by EdgeR are also found in DESeq2, and DESE2 identifies an additional set of DE genes for each experiment. 

The WT/Nr experiment invovles overexpression of nog to block regeneration. Can therefore use DE of nog as a type of sanity check to confirm identification process is working and as a way to orient FC to samples (i.e., nog should be Up in Nr). 
Using EdgeR this works fine. 

Using DESeq2 I observe that nog.L is upregulated in NR, and nog.S is upregulated in WT. 
Looking at raw counts for these genes:
nog.L counts: wt1-3 = 84, 90, 198; Nr1-3 = 1922, 2704, 35
nog.S counts: wt1-3 = 71, 71, 44; Nr1-3 = 25, 22, 5

EdgeR has identified nog.L as overexpressed in Nr, but does not classify nog.S as DE. 










